src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java

```diff
@@ -208,6 +215,40 @@ public boolean accept(Path path) {
     }
 
     /**
+     * Ensure reads in a pair fall in the same partition (input split), if the reads are queryname-sorted,
+     * so they are processed together. No shuffle is needed.
+     */
+    JavaRDD<GATKRead> putPairsInSamePartition(final SAMFileHeader header, final JavaRDD<GATKRead> reads) {
```

Comment:
Can you explain again why we're moving this logic into GATK4? It really would be better to have it part of Hadoop-BAM, since it's unlikely a client would want to load queryname-sorted reads without this invariant.